### PR TITLE
HPCC-32345 Doc: How to contribute to platform documentation

### PR DESCRIPTION
**Pull request by JamesDeFabia**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: devdoc/docs/ContributeDocs.md
Comment (stuartort) [inline devdoc/docs/ContributeDocs.md:21]: The guidelines described in this file will apply to be community and non-community.  This will be published into our public github report, but I suggest removing "We are an open-source project.   We wouldn't want internal people to think this only applies to external docs.

Maybe something like ... "Contributions to the platform are always welcome, and we strongly encourage developers and users to contribute documenation"

File: devdoc/docs/ContributeDocs.md
Comment (stuartort) [inline devdoc/docs/ContributeDocs.md:38]: If there is something special that we need to do in VS Code to enable the MD editing, mention it.  If there isn't then  ignore this comment

File: devdoc/docs/ContributeDocs.md
Comment (stuartort) [inline devdoc/docs/ContributeDocs.md:126]: do not reference open-source community

File: devdoc/docs/ContributeDocs.md
Comment (stuartort) [inline devdoc/docs/ContributeDocs.md:127]: can HPCC Style Guide be a hyper-link

**Review by stuartort — CHANGES_REQUESTED** [review]
This writeup talks about md files which is great.  Doesn't these guidelines also apply to the platform "books".  There should be a section about that (could be in a separate jira / pull request)

It would be helpful if specific examples can be provided so users can see your guidelines in action (picture worth 1000 words as you said).  This can be in a separate jira / pull request if desired

File: devdoc/docs/ContributeDocs.md
Comment (JamesDeFabia) [reply devdoc/docs/ContributeDocs.md:38]: @stuartort Good question. MD support is built-in. There is an extension for better Mermaid diagram preview, but that is an advanced topic to consider later.

File: devdoc/docs/ContributeDocs.md
Comment (ghalliday) [inline devdoc/docs/ContributeDocs.md:155]: How does this differ from `docs`?

File: devdoc/docs/ContributeDocs.md
Comment (ghalliday) [inline devdoc/docs/ContributeDocs.md:153]: Is this meant to be `docs`?  This document about documentation is in `devdoc/docs`

File: devdoc/docs/ContributeDocs.md
Comment (ghalliday) [inline devdoc/docs/ContributeDocs.md:176]: a large project -> change?

Why is a separate PR better?  Doesn't that make it less likely the documentation will occur.

File: devdoc/docs/ContributeDocs.md
Comment (ghalliday) [inline devdoc/docs/ContributeDocs.md:179]: This should probably be a separate document.  It is describing the craft of writing documentation, rather than what documentation should be written, and where you find the different kinds of documentation.

File: devdoc/docs/ContributeDocs.md
Comment (ghalliday) [inline devdoc/docs/ContributeDocs.md:234]: Is that the name of the platform, or the legal organisation?

Is it "HPCC Systems(R)" or "the HPCC Systems(R) platform"?  The former does not read well - particularly because it uses a plural noun for a singular object.

File: devdoc/docs/ContributeDocs.md
Comment (ghalliday) [inline devdoc/docs/ContributeDocs.md:355]: picky: (since this document is being picky) what is the reason for this?  It is not a mixture of the two it is two alternatives.

**Review by ghalliday — COMMENTED** [review]
@JamesDeFabia thanks.

A few overall comments.  I would like extra guidance about what kind of changes I might make as a developer, and where I should document them.  For instance

* Adding a new option to a existing language statement (user centric)
* Adding a new language statement (user centric, but possibly developer as well)
* Allowing soapcalls to be configured in a new way (targeted to operations and programmers)
* Changing the way that the dynamic file spilling works (details only for other developers)
* Extending the tests in the regression suite
* Adding design documentation for a new feature
* Discussing new configuration settings in the hel chart (and how they relate to bare-metal).

I suspect there are many more, but those are a few examples.

The other area it would be good to clarify is where is the boundary between documentation and developer's responsibilities.  Should developers do everything, and documentation only comment.  Should developers provide the skeleton information, and documentation then take that and clean it up in a separate PR?  In the same PR?  How do we stop issues with style (not most developer's strong point)  delaying features being added?
Does the process differ for the different types of documentation?

File: devdoc/docs/ContributeDocs.md
Comment (JamesDeFabia) [reply devdoc/docs/ContributeDocs.md:67]: Good point. I have been looking at tools that automatically create a step-by-step guide with screenshots that a developer could use. One good one (but it is not free) is Scribehow.com (blocked on managed machine). 
I will reword this to say the doc team can produce the screen caps later. The developer can (and should) help provide the steps to the doc team.

File: devdoc/docs/ContributeDocs.md
Comment (JamesDeFabia) [reply devdoc/docs/ContributeDocs.md:87]: Some LRM topics are separated like that (#OPTION, for example)  This is in the optional section because it is not always appropriate.

File: devdoc/docs/ContributeDocs.md
Comment (JamesDeFabia) [reply devdoc/docs/ContributeDocs.md:142]: Good question. I will reword to make this clearer

File: devdoc/docs/ContributeDocs.md
Comment (JamesDeFabia) [reply devdoc/docs/ContributeDocs.md:151]: typo. devdoc is fine. My vision is a place for higher-level info and a catch-all for anything that doesn't belong near the code.

File: devdoc/docs/ContributeDocs.md
Comment (JamesDeFabia) [reply devdoc/docs/ContributeDocs.md:153]: yes, but i am open to suggestion for a better name for this folder

File: devdoc/docs/ContributeDocs.md
Comment (JamesDeFabia) [reply devdoc/docs/ContributeDocs.md:176]: I envisioned separate PRs so the doc doesn't delay the code or the code review. But, of course, that is your call, not mine :)

File: devdoc/docs/ContributeDocs.md
Comment (JamesDeFabia) [reply devdoc/docs/ContributeDocs.md:179]: I had originally had it as a separate doc. I will do that again

File: devdoc/docs/ContributeDocs.md
Comment (JamesDeFabia) [reply devdoc/docs/ContributeDocs.md:234]: Will clarify. I agree with the plural organization name being awkward in text and conversation.

File: devdoc/docs/ContributeDocs.md
Comment (JamesDeFabia) [reply devdoc/docs/ContributeDocs.md:355]: I chose blue-green over blue/green because it wss the most common usage I found and I had to pick one for the sake of consistency. If you prefer blue/green I am ok with either. 
In fact, now that you had me take a closer look, I think blue/green **is** better.

**Review by ghalliday — COMMENTED** [review]
@JamesDeFabia do you have responses to my overall comment from the previous review?

**Comment by JamesDeFabia** [comment]
@ghalliday I missed reading part of your comments. I will add more examples to better explain different scenarios. 
I may need some additional guidance about the boundaries between documentation and developer's responsibilities. I will discuss this with @stuartort offline.

**Comment by JamesDeFabia** [comment]
@ghalliday I believe I have addressed your comments in the latest revision. Thanks for your patience.

File: devdoc/docs/ContributeDocs.md
Comment (ghalliday) [inline devdoc/docs/ContributeDocs.md:158]: We don't do this at the moment e.g. devdoc/roxie.md.  I don't think it has any advantages.

We did introduce sourcedoc.xml files in each directory a while back, but they never caught on.

File: devdoc/docs/ContributeDocs.md
Comment (ghalliday) [inline devdoc/docs/ContributeDocs.md:169]: Currently devdoc/userdoc.  Should it move?  Should devdoc and userdoc both be under doc?

File: devdoc/docs/ContributeDocs.md
Comment (ghalliday) [inline devdoc/docs/ContributeDocs.md:191]: need to say why it is better.

a separate Pull request may allow the code change to be merged more quickly?

File: devdoc/docs/ContributeDocs.md
Comment (ghalliday) [inline devdoc/docs/ContributeDocs.md:197]: typo: paramater

File: devdoc/docs/ContributeDocs.md
Comment (ghalliday) [reply devdoc/docs/ContributeDocs.md:48]: Any thoughts on this comment?

Comment (JamesDeFabia) [reply devdoc/docs/ContributeDocs.md:48]: I do not think the LRM should be be restructured. It does contain all of these elements, just not in that order. I think creating more sample templates (similar to the one I added in another PR)  might help.

File: devdoc/docs/ContributeDocs.md
Comment (JamesDeFabia) [reply devdoc/docs/ContributeDocs.md:169]: I like the folder structure we created before I got involved. This can change it later if we choose. 
I am adding full paths to folders in the doc (e.g., **devdoc**/**userdoc**/**blogs**)

File: devdoc/docs/ContributeDocs.md
Comment (ghalliday) [inline devdoc/docs/ContributeDocs.md:153]: If you're manually splitting lines this needs joining.

**Review by ghalliday — COMMENTED** [review]
Looks close to being ready to merge.  One trivial comment.  Please squash.

**Comment by JamesDeFabia** [comment]
I made the trivial edit and squashed

**Review by stuartort — APPROVED** [review]
Looks good.  I made 1 comment which may or may not need to be addressed

File: devdoc/docs/HPCCStyleGuide.md
Comment (JamesDeFabia) [reply devdoc/docs/HPCCStyleGuide.md:119]: Modern browsers force a redirect from HTTP to HTTPS, but not vice versa. I would like to leave is as is in the doc.

**Comment by ghalliday** [comment]
Merged - good enough for a first iteration.

**Pull request merged**
